### ISSUE TITLE
In Edit, can't make out most icons on dark theme and high res screen

### ISSUE THREAD

**Issue by genachka** [shutter-project/shutter#382]
Using 0.98, though this was an issue prior versions too. My specs are:
* Ubuntu 21.04
* Gnome 3.38.5
* Screen res: 3840 x 2160 with Font scaling factor 1.25
* Using Yaru-dark theme

The issue: 
The left sidebar icons are almost impossible to see both because of their size and lack of clarity in the dark theme. The top toolbar icons are a little better on the left (like close, save) but then to the right of those it becomes hard to make them out for same reason.

![diI3VyLuI3](https://user-images.githubusercontent.com/7947294/128362578-1b1ec35f-a170-43c4-8e10-6310c0b8270e.png)

**Comment by Photon89** [comment]
The icons in the top bar are loaded from the icon theme currently in use, maybe a different icon theme would help? Though the contrast seems fine to me, it's merely about the icon size, I guess. How do other apps perform here? Maybe you could use HiDPI scaling to make everything larger? As far as I understand, an app cannot control the icon size depending on the resolution, and for non-HiDPI resolutions the icon size is just fine.

The icons on the left panel are hard coded Tango icons and some of them indeed are problematic on dark themes. Ideally, we could find suiting icons which are supplied by the icon themes and fit the GTK theme.

**Comment by genachka** [comment]
If the top icon bar icons were same size that Close, Save are it would be consistent and easier to see. I don't have an issue in any other app so increasing scaling or enabling HiDPI scaling globally just to make shutter work would cause those other apps to have overly large icons.

Though many of the apps offer their own scaling options to increase so maybe this option can be added here in preferences.

But agree, the bigger issue is the left sidebar icons. It would be helpful to allow setting additional scaling options in preferences (1 setting for all icons that Shutter uses) as they are tiny and finding icons that work better on darker themes.

**Comment by Photon89** [comment]
> If the top icon bar icons were same size that Close, Save are it would be consistent and easier to see.

It looks like a theme specific issue to me. Here is how it looks like on my end with the Delft theme:

![Bildschirmfoto_2021-08-05_16-21-19](https://user-images.githubusercontent.com/1191480/128366301-0152cc33-63bd-4b6f-9714-6d56f2d08470.png)

I think, all the icons have about the same size here.

**Comment by genachka** [comment]
So if there's anything that can be done to get it to work better with dark themes, like the ones that come with Ubuntu that I'm using it would be helpful, especially with the left side icons. I'm happy to see Shutter getting some love again as for the past couple years I've been forced to switch to something like Flameshot which is getting better but doesn't have many of the features I've grown to love in Shutter. But as you can at least see in the screenshot, it has no issues with icons which are well defined and easy to see with my existing resolution and dark theme. I'm not straining or trying to guess what they are / mean.


![F7rsvohWb8](https://user-images.githubusercontent.com/7947294/128368889-f17d5d6b-b74c-402d-81f7-1369e166c7e2.png)

**Comment by Photon89** [comment]
I see your point, those are really very clear. But they appear in the selection screen rather than inside a window and thus are less disturbing. Maybe it's just me but I'd find it not really aesthetic to have flat violet icons in a GTK GUI, though they are clearly very functional. 

I search how other apps handle this issue:

- Gimp has just a dark interface with white icons, regardless of the actual GTK theme. It's probably hard to implement and doesn't fit into the desktop nicely.
- Inkscape has really many GUI options (choice of GTK theme, icon theme, a checkbox for using symbolic icons etc). Looks like the best solution but probably rather hard to implement.
- Xournalpp has a checkbox "Use dark theme" which loads a second icon set with light icons. 

Ideally, one could find icons which work well on both dark and light themes. Basically what Flameshot does, but not in violet such that it integrates better into a GTK GUI. Not sure if such icons exist though...

**Comment by DarthGandalf** [comment]
> Gimp has just a dark interface with white icons, regardless of the actual GTK theme. It's probably hard to implement and doesn't fit into the desktop nicely.

I think it should be easy to implement via CSS which gtk supports. But I agree wrt desktop. OTOH I don't find gtk nice either way: any Qt app I saw looks so much better than any Gtk app I saw. But I don't think users of light themes would be happy if we make Shutter even uglier than it is now by forcing dark theme on everyone, which as you said, doesn't fit

> Inkscape has really many GUI options (choice of GTK theme, icon theme, a checkbox for using symbolic icons etc). Looks like the best solution but probably rather hard to implement.

What's symbolic icons?

> Ideally, one could find icons which work well on both dark and light themes. Basically what Flameshot does

From what I can see in https://github.com/flameshot-org/flameshot/tree/master/data/img/material they used material icons, and made them in 2 colors: white or black.

**Comment by Photon89** [comment]
> What's symbolic icons?

Flat black/white icons, see screenshot from Inkscape:

![Auswahl_003](https://user-images.githubusercontent.com/1191480/128519313-95d9d102-b20d-4dea-9ba5-3cd218ba7c38.png)

> From what I can see in https://github.com/flameshot-org/flameshot/tree/master/data/img/material they used material icons, and made them in 2 colors: white or black.

But how did the icons become violet? Probably, the violet circles are not part of the actual icons but are added programmatically?

**Comment by DarthGandalf** [comment]
> Probably, the violet circles are not part of the actual icons but are added programmatically?

Yep.

**Comment by genachka** [comment]
Saw some work on the 0.99 release mention changes to the editor's icons thank you but it's still an issue on dark theme and higher resolutions. Still can't make them out. :( 

There's two issue here:
1. One is the size of the icons in high resolution is making them very hard to see. 
2. The icon coloring in dark doesn't work well, but maybe with 0.99 if the size of the icons were better scaled for higher resolutions, especially the left side-bar icons then most of the icons would probably be better understandable with exception of these which are black on black: `Draw a straight line`, `Draw an arrow`, `Add an auto-increment shape to the screenshot`

Here's a side by side of what Shutter 0.99 icons (labeled as #1 in screenshot below) and Inkscape (labeled as #2 in screenshot below) icons look like on my dark theme and resolution. I can make out each and every Inkscape icon easily -- they are larger and coloring stands out well.
![2021-09-08_10-08](https://user-images.githubusercontent.com/7947294/132525541-fac80360-6a08-4475-ad71-27a3587138e1.png)

**Comment by Photon89** [comment]
@genachka Unfortunately, those changes have no visible effect at all.

@DarthGandalf We could actually make the icons in the toolbar on the left larger than 24px (32px maybe?). 

Problem is, they are PNGs and I don't know if the SVG sources or just larger versions are available somewhere. @vadi2 Maybe you know something?

**Comment by cowabungaa** [comment]
I don't use the edit much but the rare time that i do it was always difficult to see which one was the arrow and the line. i inverted the color of some icons for a quick fix. i tried importing a bigger 50px png but it was still scaled down. i don't know if this will help you, i can work with this until the toolbar gets updated
![2021-09-08-143250](https://user-images.githubusercontent.com/23160213/132590986-a4645d9d-1cf2-4814-bc64-5ced30969d05.png)

edit: /usr/share/shutter/resources/icons/drawing_tool - if u want to style your own icons and come back to shutter

**Comment by tethragon** [comment]
> I don't use the edit much but the rare time that i do it was always difficult to see which one was the arrow and the line. i inverted the color of some icons for a quick fix. i tried importing a bigger 50px png but it was still scaled down. i don't know if this will help you, i can work with this until the toolbar gets updated ![2021-09-08-143250](https://user-images.githubusercontent.com/23160213/132590986-a4645d9d-1cf2-4814-bc64-5ced30969d05.png)
> 
> edit: /usr/share/shutter/resources/icons/drawing_tool - if u want to style your own icons and come back to shutter

Thank you for this wonderful idea! At least I can use shutter now until the devs think a solution for dark themes.
Here are the icons I replaced (for dark themes). I inverted the colors mainly and I chose a different icon for censor

![draw-arrow](https://user-images.githubusercontent.com/39238674/231162325-087090e5-e321-4b70-8162-8641fce752d3.png)

![draw-line](https://user-images.githubusercontent.com/39238674/231162499-3b265965-1dcc-457a-b67b-cdca9f8987b6.png)

![draw-number](https://user-images.githubusercontent.com/39238674/231162552-49646202-4077-4f8e-8270-96a32f07246d.png)

![draw-text](https://user-images.githubusercontent.com/39238674/231162605-62e008b9-4314-4006-8513-1e6b3db7c511.png)

![draw-censor](https://user-images.githubusercontent.com/39238674/231162681-694d0719-c06d-44ae-8193-eef865cf20a1.png)

And now my icon row looks like this:

![Screenshot from 2023-04-11 15-31-08](https://user-images.githubusercontent.com/39238674/231163002-47f0ba5d-38a5-459d-b751-eb56402a1fca.png)

At least now I can use shutter properly!

**Comment by pmorch** [comment]
@cowabungaa  and @tethragon : What exactly did you do? Did you manually replace the individual icons under `/usr/share/shutter/resources/icons/drawing_tool`? Could you link to a .zip file or something with all the updated icons?

**Comment by cowabungaa** [comment]
hey there, looks like there were some changes since then :) the steps in my post no longer work 

these steps work for me (v0.99.6)

1. git clone repo / download release (skip if already done)
2. cd shutter-0.99.6/share/shutter/resources/icons/drawing_tool
3. tethragon provided some icons in the post. download and replace the icons in the folder. or open the icons in graphics editor and edit however you like
4. cd shutter-0.99.6/ && sudo make install
5. :)

**Comment by pmorch** [comment]
Another thing that can be done is to run shutter specifically with another (e.g. light) theme. Like this from the command line:

```shell
$ GTK_THEME=Adwaita shutter
```

Or modify a `.desktop` file like this:
```
Exec=env GTK_THEME=Adwaita shutter %u
```

it will of course appear light (duh), but the icons don't look broken.

**Comment by cowabungaa** [comment]
you're right

but the issue is that the icons are hard to make out on dark themes :/

**Comment by tethragon** [comment]
I've created and refined a set of custom icons for the Shutter drawing tool that dramatically improves their appearance on dark themes.

<img width="44" height="453" alt="Image" src="https://github.com/user-attachments/assets/e6def21d-09b5-45dc-85cd-e58c2734fb02" />

If you like this workaround fix:

1) Download the icons:
[shutter editor icons for dark themes.zip](https://github.com/user-attachments/files/24014164/shutter.editor.icons.for.dark.themes.zip)

2) Replace the existing icons in this path: /usr/share/shutter/resources/icons/drawing_tool

Hopefully, the Shutter developers will integrate a native solution for dark theme icon appearance in a future update!

**Comment by Photon89** [comment]
@DarthGandalf I'd propose to add a checkbox to the settings which switches between icons for light and dark GTK themes. Question is: Where to put this checkbox? What about Main category (make a new section "Drawing Tool" and add a checkbox along the lines "Use light icons for dark backgrounds")?

**Comment by DarthGandalf** [comment]
Such checkbox shouldn't exist - instead, shutter should pick up from the system whether the current theme is light or dark, and pick the correct icons.

If you want it to be configurable, there should be 3 options - light, dark, system.

I'm not sure how to do that with GTK though. I can query the name of the theme using D-Bus, but then need to parse the name???

```
$ gsettings get org.gnome.desktop.interface gtk-theme
'Orchis-Light'
```

From what I can see in https://wiki.archlinux.org/title/Dark_mode_switching#GTK often the name contains just "Adwaita"

**Comment by Photon89** [comment]
Of course, this would be even better, but I'm not sure whether it is possible at all. The non-custom icons can be changed system-wide (not just for Shutter) through the DEs appearance settings config tool. The icon theme is independent of the GTK theme, so there is no automatic switching globally either. The user is in charge of choosing an icon theme which suits his GTK theme.

In Shutter's Drawing Tool we use custom icons so we cannot rely on the user's icon theme choice. But same as in the case of the global icon theme the user needs to choose the custom icons manually. For consistency reasons but also because it seems technically impossible to automate the icon choice.

> If you want it to be configurable, there should be 3 options - light, dark, system.

But we have no system icon set, just two sets, for light and dark backgrounds.

**Comment by DarthGandalf** [comment]
> But we have no system icon set, just two sets, for light and dark backgrounds.

No, the system would be the one which queries the system settings and chooses light or dark automatically

This is what chatgpt said:

```
GTK 3: Detect dark theme via style context (heuristic)

GTK 3 does not expose a reliable “dark theme” flag. You must infer it.

Example
GtkStyleContext *context =
    gtk_widget_get_style_context(widget);

GdkRGBA bg;
gtk_style_context_get_background_color(
    context,
    GTK_STATE_FLAG_NORMAL,
    &bg
);

gboolean is_dark =
    (bg.red + bg.green + bg.blue) / 3.0 < 0.5;


This is fragile and not recommended.
```

I haven't tested this, but it might work?

**Comment by Photon89** [comment]
Good idea, I will try to have a look into it tomorrow if time allows. For the moment i pushed a draft PR which only works with a fixed variable for now.

**Comment by Photon89** [comment]
> No, the system would be the one which queries the system settings and chooses light or dark automatically

I'll try to implement this. Where do we put the checkbox though?

**Comment by DarthGandalf** [comment]
Dunno, as a radio button in one of the setting sections?
Probably "system" should be named "auto" instead.

Another point: sometimes the theme can change dynamically, e.g. dark theme at night, light theme at day, so this shouldn't be set once on startup.

**Comment by Photon89** [comment]
> Dunno, as a radio button in one of the setting sections?

Main section in a new subsection "Drawing Tool" works for you?

**Comment by DarthGandalf** [comment]
Sure

**Comment by Photon89** [comment]
@tethragon Thanks for the adjusted icons!

A prototype can be tested in latest git.
